### PR TITLE
Remove the blanket restriction against nested usage.

### PR DESCRIPTION
**Pull request by mikewest**
Some credential types wish to be usable in nested contexts. This patch moves the
top-level restriction out of the 'navigator.credentials' API itself, and into the
credential types that wish to have such a restriction.

Note that it also changes the behavior of those restrictions: rather than throwing
a 'NotSupported' error, it instead returns an empty set. That seems both simpler and
less likely to cause developers to be angry at me.

***
[Preview](https://s3.amazonaws.com/pr-preview/w3c/webappsec-credential-management/top-level.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/webappsec-credential-management/ac0ea41...4bb98cc.html)

### CODE REVIEW

**Comment by mikewest** [comment]
@battre: WDYT? This should address #113 (and, by extension, w3c/webauthn#374). I'll put some tests together in a separate patch.

/cc @equalsJeffH

File: index.src.html [hunk 1]
```diff
@@ -1810,11 +1832,19 @@ <h4 algorithm id="construct-federatedcredential-data">
   granting access to credentials for an origin other than the <a>top-level browsing context</a>,
   which is the only security origin which users can reasonably be expected to understand.
 
-  Therefore, both {{get()}} and {{store()}} will result in a rejected {{Promise}} when called from
-  a [=nested browsing context=], and the API and related interfaces are not exposed inside
-  {{Worker}} contexts.
+  This document exposes the Credential Management APIs to those contexts, as it's likely that some
+  credential types will be straightforward to make available if user agents put enough thought and
```
Comment (battre) [inline index.src.html]: Is "to make available" grammatically correct?

File: index.src.html [hunk 2]
```diff
@@ -695,15 +695,7 @@ <h4 id="algorithm-request" algorithm>Request a `Credential`</h4>
 
     2.  Assert: |settings| is a [=secure context=].
 
-    3.  Return [=a promise rejected with=] `NotSupportedError` if any of the following statements
-        are true:
-
-        1.  |settings| does not have a [=environment settings object/responsible document=].
-
-        2.  |settings|' [=environment settings object/responsible document=] is not the
```
Comment (battre) [inline index.src.html]: This check is now moved into the specific implementation that are executed `[=in parallel=]`. Is that ok? There is a pending pull request to limit access to the `[=environment settings object=]` in parallel execution.

File: index.src.html [hunk 3]
```diff
@@ -1173,7 +1154,18 @@ <h4 algorithm id="collectfromcredentialstore-passwordcredential">
...
-    2.  If |options|["{{CredentialRequestOptions/password}}"] is not `true`, return the empty set.
+    2.  Return the empty set if any of the following are true:
+
+        1.  |options|["{{CredentialRequestOptions/password}}"] is not `true`.
+
+        2.  The [=current settings object=] does not have a
+            [=environment settings object/responsible document=].
+
+        3.  The [=current settings object=]'s [=environment settings object/responsible document=]
+            is not the [=active document=] in a [=top-level browsing context=].
```
Comment (battre) [inline index.src.html]: I stumbled about the "**a** [=top-level browsing context=]". Is that correct? Should this be "the" with a qualified top-level browsing context?

File: index.src.html (hunk 1 above)
Comment (mikewest) [reply index.src.html]: It sounds fine to my ear, but I'm happy to reword it to something like "... as it's likely that some credential types will be straightforward to make available _in a cross-origin nested context_ if user agents put enough thought and..." if that would be helpful.

File: index.src.html (hunk 3 above)
Comment (mikewest) [reply index.src.html]: The browser can have multiple top-level browsing contexts, so "the" seems overly restrictive. *shrug* Looks like the HTML spec uses "a top-level browsing context" 42 times, and "the top-level browsing context" 27 times. I think it's fine either way. :)

File: index.src.html (hunk 2 above)
Comment (mikewest) [reply index.src.html]: Hrm. I think you're right. We're now calling `current settings object`, which I guess is difficult to get once we're inside the parallel world. That's annoying.

I suppose we could pass in an "is same-origin with the top-level document" flag instead. That's a little annoying, but whatever.

File: index.src.html (hunk 1 above)
Comment (battre) [reply index.src.html]: @equalsJeffH Do you agree with this proposal? It would be a change to WebAuthN but would not have any impact on the implementations, I guess.

File: index.src.html (hunk 2 above)
Comment (battre) [reply index.src.html]: @equalsJeffH Do you agree with this proposal? It would be a change to WebAuthN but would not have any impact on the implementations, I guess.

File: index.src.html (hunk 3 above)
Comment (battre) [reply index.src.html]: About the multiple top-level browsing contexts.

What would happen if 

* www.example.com is opened in a tab
* www.foobar.com is opened in another tab with an iframe for www.example.com, in that iframe, www.example.com tries to retrieve a password credential.

Would the top-level browsing context of the first tab enable the second access?

Comment (mikewest) [reply index.src.html]: This seems like it would work as expected: the framed `example.com` would look at its own settings object's responsible document, and note that it is not the active document in a top-level browsing context (because it's in a frame).

The fact that `example.com` is open in another window doesn't change the state of the context actually calling the method.

File: index.src.html (hunk 1 above)
Comment (mikewest) [reply index.src.html]: The appeal to authority fallacy wins again! :)

File: index.src.html (hunk 2 above)
Comment (mikewest) [reply index.src.html]: The change to WebAuthN would basically be accepting a new boolean and throwing it away. Seems reasonable enough to me, if it enables behavior that WebAuthN values.

**Comment by mikewest** [comment]
@battre: Mind taking another look? This changes the behavior a little bit more (in ways that I think we've discussed before), by allowing same-origin frames direct access (as the UI problems there should be minimal). It also take the opportunity to resolve w3c/webappsec-credential-management#98 as a side-effect of poking at `[[Store]]()`.

WDYT?

**Comment by mikewest** [comment]
Looking at the implementation and tests, I think this will actually be simpler for developers to understand if we keep the exception in place (but shift to `NotAllowedError` rather than `NotSupportedError`). Tests to follow shortly.

File: index.src.html (hunk 3 above)
Comment (battre) [reply index.src.html]: I had interpreted "is not the [=active document=] in a [=top-level browsing context=]" as "is not the [=active document=] in **any** [=top-level browsing context=]", where the latter www.example.com would be just another top-level browsing context.

**Comment by equalsJeffH** [comment]
Am catching up here -- I was at IETF-100 when this was created and eventually merged.  This seems change seems fine to me. It _may_ be declared sufficient to close w3c/webauthn#374, I've inquired as to that: https://github.com/w3c/webauthn/issues/374#issuecomment-347632509
